File: desktop-windows/src/core/signaling_server.rs
//! 内置 WebSocket 信令服务器（tokio-tungstenite）。
//! 与 `signaling-server/src/index.js` 协议完全一致，零外部依赖，用于局域网内置模式。
//!
//! 协议（JSON 文本帧）：
//!   C->S {type:'create-room'}                  -> S->C {type:'room-created', roomId, token}
//!   C->S {type:'join-room', roomId, token}     -> 双方: {type:'peer-joined', role}
//!   C->S {type:'signal', data}                 -> 原样转发给房间内另一端
//!   C->S {type:'message', data}                -> 业务消息中继给房间内另一端
//!   S->C {type:'peer-left'}                     某端断开时通知另一端

use crate::core::wire;
use futures_util::{SinkExt, StreamExt};
use rand::RngCore;
use serde_json::{json, Value};
use std::collections::HashMap;
use std::sync::Arc;
use tokio::net::TcpListener;
use tokio::sync::{mpsc, Mutex};
use tokio_tungstenite::tungstenite::Message;

/// 服务器状态/日志事件。
pub enum ServerEvent {
    /// running, port
    State(bool, u16),
    Error(String),
    Log(String),
}

type ClientId = u64;
type ClientSink = mpsc::UnboundedSender<Message>;

#[derive(Default)]
struct Room {
    token: String,
    host: Option<ClientId>,
    guest: Option<ClientId>,
}

#[derive(Default)]
struct Shared {
    clients: HashMap<ClientId, ClientSink>,
    client_room: HashMap<ClientId, String>,
    rooms: HashMap<String, Room>,
}

/// 内置信令服务器。`start` 在指定端口监听，`stop` 关闭。
pub struct SignalingServer {
    events: mpsc::UnboundedSender<ServerEvent>,
    shutdown: Arc<Mutex<Option<tokio::sync::oneshot::Sender<()>>>>,
    shared: Arc<Mutex<Shared>>,
    next_id: Arc<std::sync::atomic::AtomicU64>,
}

impl SignalingServer {
    pub fn new(events: mpsc::UnboundedSender<ServerEvent>) -> Self {
        SignalingServer {
            events,
            shutdown: Arc::new(Mutex::new(None)),
            shared: Arc::new(Mutex::new(Shared::default())),
            next_id: Arc::new(std::sync::atomic::AtomicU64::new(1)),
        }
    }

    /// 在指定端口启动。若已在运行会先停止。
    pub async fn start(&self, port: u16) {
        self.stop().await;

        let listener = match TcpListener::bind(("0.0.0.0", port)).await {
            Ok(l) => l,
            Err(_) => {
                let _ = self
                    .events
                    .send(ServerEvent::Error(format!("端口 {} 监听失败（可能被占用）", port)));
                let _ = self.events.send(ServerEvent::State(false, 0));
                return;
            }
        };

        let (tx_shutdown, mut rx_shutdown) = tokio::sync::oneshot::channel::<()>();
        *self.shutdown.lock().await = Some(tx_shutdown);

        let _ = self.events.send(ServerEvent::State(true, port));
        let _ = self
            .events
            .send(ServerEvent::Log(format!("信令服务器已启动 ws://0.0.0.0:{}", port)));

        let shared = self.shared.clone();
        let next_id = self.next_id.clone();
        let events = self.events.clone();

        tokio::spawn(async move {
            loop {
                tokio::select! {
                    _ = &mut rx_shutdown => {
                        break;
                    }
                    accepted = listener.accept() => {
                        match accepted {
                            Ok((stream, _addr)) => {
                                let id = next_id.fetch_add(1, std::sync::atomic::Ordering::SeqCst);
                                let shared = shared.clone();
                                tokio::spawn(handle_connection(stream, id, shared));
                            }
                            Err(_) => break,
                        }
                    }
                }
            }
            // 关闭：清理所有客户端与房间。
            let mut g = shared.lock().await;
            g.clients.clear();
            g.client_room.clear();
            g.rooms.clear();
            let _ = events.send(ServerEvent::Log("信令服务器已停止".to_string()));
            let _ = events.send(ServerEvent::State(false, 0));
        });
    }

    pub async fn stop(&self) {
        if let Some(tx) = self.shutdown.lock().await.take() {
            let _ = tx.send(());
        }
    }
}

async fn handle_connection(
    stream: tokio::net::TcpStream,
    id: ClientId,
    shared: Arc<Mutex<Shared>>,
) {
    let ws = match tokio_tungstenite::accept_async(stream).await {
        Ok(ws) => ws,
        Err(_) => return,
    };
    let (mut write, mut read) = ws.split();
    let (tx, mut rx) = mpsc::unbounded_channel::<Message>();

    shared.lock().await.clients.insert(id, tx);

    // 发送任务。
    let send_task = tokio::spawn(async move {
        while let Some(msg) = rx.recv().await {
            if write.send(msg).await.is_err() {
                break;
            }
        }
    });

    while let Some(item) = read.next().await {
        match item {
            Ok(Message::Text(text)) => {
                handle_message(&text, id, &shared).await;
            }
            Ok(Message::Binary(bin)) => {
                if let Ok(text) = String::from_utf8(bin) {
                    handle_message(&text, id, &shared).await;
                }
            }
            Ok(Message::Close(_)) | Err(_) => break,
            _ => {}
        }
    }

    cleanup(id, &shared).await;
    send_task.abort();
}

async fn handle_message(text: &str, id: ClientId, shared: &Arc<Mutex<Shared>>) {
    let obj: Value = match serde_json::from_str(text) {
        Ok(v) => v,
        Err(_) => {
            send_to(shared, id, &json!({"type": wire::signal::ERROR, "reason": wire::reason::INVALID_JSON})).await;
            return;
        }
    };
    let msg_type = obj.get(wire::key::TYPE).and_then(|v| v.as_str()).unwrap_or("");

    match msg_type {
        t if t == wire::signal::CREATE_ROOM => {
            let room_id = gen_id(3);
            let token = gen_id(8);
            {
                let mut g = shared.lock().await;
                g.rooms.insert(
                    room_id.clone(),
                    Room {
                        token: token.clone(),
                        host: Some(id),
                        guest: None,
                    },
                );
                g.client_room.insert(id, room_id.clone());
            }
            send_to(
                shared,
                id,
                &json!({"type": wire::signal::ROOM_CREATED, "roomId": room_id, "token": token}),
            )
            .await;
        }

        t if t == wire::signal::JOIN_ROOM => {
            let room_id = obj.get("roomId").and_then(|v| v.as_str()).unwrap_or("").to_string();
            let token = obj.get("token").and_then(|v| v.as_str()).unwrap_or("").to_string();
            let mut host_id: Option<ClientId> = None;
            let result_reason = {
                let mut g = shared.lock().await;
                match g.rooms.get_mut(&room_id) {
                    None => Some(wire::reason::ROOM_NOT_FOUND),
                    Some(room) if room.token != token => Some(wire::reason::BAD_TOKEN),
                    Some(room) if room.guest.is_some() => Some(wire::reason::ROOM_FULL),
                    Some(room) => {
                        room.guest = Some(id);
                        host_id = room.host;
                        None
                    }
                }
            };
            match result_reason {
                Some(reason) => {
                    send_to(shared, id, &json!({"type": wire::signal::ERROR, "reason": reason})).await;
                }
                None => {
                    shared.lock().await.client_room.insert(id, room_id.clone());
                    send_to(shared, id, &json!({"type": wire::signal::PEER_JOINED, "role": wire::role::GUEST})).await;
                    if let Some(hid) = host_id {
                        send_to(shared, hid, &json!({"type": wire::signal::PEER_JOINED, "role": wire::role::HOST})).await;
                    }
                }
            }
        }

        t if t == wire::signal::SIGNAL => {
            if let Some(peer) = other_peer(shared, id).await {
                if let Some(data) = obj.get("data") {
                    send_to(shared, peer, &json!({"type": wire::signal::SIGNAL, "data": data})).await;
                }
            } else {
                send_to(shared, id, &json!({"type": wire::signal::ERROR, "reason": wire::reason::NOT_IN_ROOM})).await;
            }
        }

        t if t == wire::signal::MESSAGE => {
            if let Some(peer) = other_peer(shared, id).await {
                if let Some(data) = obj.get(wire::key::DATA) {
                    send_to(shared, peer, &json!({wire::key::TYPE: wire::signal::MESSAGE, wire::key::DATA: data})).await;
                }
            } else {
                send_to(shared, id, &json!({"type": wire::signal::ERROR, "reason": wire::reason::NOT_IN_ROOM})).await;
            }
        }

        _ => {
            send_to(shared, id, &json!({"type": wire::signal::ERROR, "reason": wire::reason::UNKNOWN_TYPE})).await;
        }
    }
}

async fn other_peer(shared: &Arc<Mutex<Shared>>, id: ClientId) -> Option<ClientId> {
    let g = shared.lock().await;
    let room_id = g.client_room.get(&id)?;
    let room = g.rooms.get(room_id)?;
    if room.host == Some(id) {
        room.guest
    } else if room.guest == Some(id) {
        room.host
    } else {
        None
    }
}

async fn cleanup(id: ClientId, shared: &Arc<Mutex<Shared>>) {
    let peer = other_peer(shared, id).await;
    {
        let mut g = shared.lock().await;
        g.clients.remove(&id);
        if let Some(room_id) = g.client_room.remove(&id) {
            let mut remove_room = false;
            if let Some(room) = g.rooms.get_mut(&room_id) {
                if room.host == Some(id) {
                    room.host = None;
                }
                if room.guest == Some(id) {
                    room.guest = None;
                }
                remove_room = room.host.is_none() && room.guest.is_none();
            }
            if remove_room {
                g.rooms.remove(&room_id);
            }
        }
    }
    if let Some(peer) = peer {
        send_to(shared, peer, &json!({"type": wire::signal::PEER_LEFT})).await;
    }
}

async fn send_to(shared: &Arc<Mutex<Shared>>, id: ClientId, obj: &Value) {
    let sink = {
        let g = shared.lock().await;
        g.clients.get(&id).cloned()
    };
    if let Some(sink) = sink {
        let _ = sink.send(Message::Text(obj.to_string()));
    }
}

fn gen_id(bytes: usize) -> String {
    let mut buf = vec![0u8; bytes];
    rand::thread_rng().fill_bytes(&mut buf);
    buf.iter().map(|b| format!("{:02x}", b)).collect()
}

File: desktop-macos/Sources/Core/SignalingServer.swift
import Foundation
import Network

/// 原生 WebSocket 信令服务器（Network.framework / NWProtocolWebSocket）。
/// 与 `signaling-server/src/index.js` 协议完全一致，零 Node 依赖，用于局域网内置模式。
///
/// 协议（JSON 文本帧）：
///   C->S {type:'create-room'}                  -> S->C {type:'room-created', roomId, token}
///   C->S {type:'join-room', roomId, token}     -> 双方: {type:'peer-joined', role}
///                                                 失败: {type:'error', reason}
///   C->S {type:'signal', data}                 -> 原样转发给房间内另一端
///   C->S {type:'message', data}                -> 业务消息中继给房间内另一端
///   S->C {type:'peer-left'}                     某端断开时通知另一端
final class SignalingServer {
    /// 房间内创建者是 host（桌面），加入者是 guest（手机）。
    private final class Room {
        let token: String
        weak var host: Client?
        weak var guest: Client?
        init(token: String) { self.token = token }
    }

    private final class Client {
        let connection: NWConnection
        var roomId: String?
        init(_ connection: NWConnection) { self.connection = connection }
    }

    var onLog: ((String) -> Void)?
    /// 状态变化回调：running, port。主线程触发。
    var onStateChange: ((Bool, UInt16) -> Void)?
    /// 启动失败回调（如端口被占用）。主线程触发。
    var onError: ((String) -> Void)?

    private(set) var isRunning = false
    private(set) var port: UInt16 = 0

    private var listener: NWListener?
    private var rooms: [String: Room] = [:]
    private var clients: [ObjectIdentifier: Client] = [:]
    private let queue = DispatchQueue(label: "com.chatput.signaling-server")

    // MARK: - 生命周期

    /// 在指定端口启动。若已在运行会先停止。
    func start(port requestedPort: UInt16) {
        queue.async { [weak self] in
            self?.startLocked(port: requestedPort)
        }
    }

    func stop() {
        queue.async { [weak self] in
            self?.stopLocked(notify: true)
        }
    }

    private func startLocked(port requestedPort: UInt16) {
        stopLocked(notify: false)

        guard let nwPort = NWEndpoint.Port(rawValue: requestedPort) else {
            emitState(running: false, port: 0)
            emitError("无效端口: \(requestedPort)")
            log("无效端口: \(requestedPort)")
            return
        }

        let params = NWParameters.tcp
        params.allowLocalEndpointReuse = true
        let wsOptions = NWProtocolWebSocket.Options()
        wsOptions.autoReplyPing = true
        params.defaultProtocolStack.applicationProtocols.insert(wsOptions, at: 0)

        do {
            let listener = try NWListener(using: params, on: nwPort)
            self.listener = listener
            self.port = requestedPort

            listener.stateUpdateHandler = { [weak self] state in
                guard let self = self else { return }
                self.queue.async {
                    switch state {
                    case .ready:
                        self.isRunning = true
                        self.emitState(running: true, port: requestedPort)
                        self.log("信令服务器已启动 ws://0.0.0.0:\(requestedPort)")
                    case .failed(let error):
                        self.log("信令服务器失败: \(error)")
                        self.emitError("端口 \(requestedPort) 监听失败（可能被占用）")
                        self.stopLocked(notify: true)
                    case .cancelled:
                        break
                    default:
                        break
                    }
                }
            }

            listener.newConnectionHandler = { [weak self] connection in
                self?.queue.async { self?.accept(connection) }
            }

            listener.start(queue: queue)
        } catch {
            log("无法监听端口 \(requestedPort): \(error)")
            emitError("无法监听端口 \(requestedPort)（可能被占用）")
            emitState(running: false, port: 0)
        }
    }

    private func stopLocked(notify: Bool) {
        for client in clients.values {
            client.connection.cancel()
        }
        clients.removeAll()
        rooms.removeAll()
        listener?.cancel()
        listener = nil
        let wasRunning = isRunning
        isRunning = false
        if notify && wasRunning {
            log("信令服务器已停止")
            emitState(running: false, port: port)
        }
        port = 0
    }

    // MARK: - 连接处理

    private func accept(_ connection: NWConnection) {
        let client = Client(connection)
        clients[ObjectIdentifier(connection)] = client

        connection.stateUpdateHandler = { [weak self] state in
            guard let self = self else { return }
            self.queue.async {
                switch state {
                case .failed, .cancelled:
                    self.cleanup(client)
                default:
                    break
                }
            }
        }
        connection.start(queue: queue)
        receive(on: client)
    }

    private func receive(on client: Client) {
        client.connection.receiveMessage { [weak self] data, context, _, error in
            guard let self = self else { return }
            self.queue.async {
                if let error = error {
                    self.log("连接接收错误: \(error)")
                    self.cleanup(client)
                    return
                }

                if let context = context,
                   let meta = context.protocolMetadata.first as? NWProtocolWebSocket.Metadata {
                    switch meta.opcode {
                    case .close:
                        self.cleanup(client)
                        return
                    case .text, .binary:
                        if let data = data, !data.isEmpty {
                            self.handle(data, from: client)
                        }
                    default:
                        break
                    }
                }
                self.receive(on: client)
            }
        }
    }

    // MARK: - 协议

    private func handle(_ data: Data, from client: Client) {
        guard let obj = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any],
              let type = obj["type"] as? String else {
            sendJSON(["type": Wire.Signal.error, "reason": Wire.Reason.invalidJSON], to: client)
            return
        }

        switch type {
        case Wire.Signal.createRoom:
            let roomId = Self.genId(bytes: 3)   // 短，便于二维码
            let token = Self.genId(bytes: 8)
            let room = Room(token: token)
            room.host = client
            rooms[roomId] = room
            client.roomId = roomId
            sendJSON(["type": Wire.Signal.roomCreated, "roomId": roomId, "token": token], to: client)

        case Wire.Signal.restoreRoom:
            let roomId = obj["roomId"] as? String ?? Self.genId(bytes: 3)
            let token = obj["token"] as? String ?? Self.genId(bytes: 8)
            // If room with this ID already exists (stale), remove it first
            if let existing = rooms[roomId] {
                if let host = existing.host { host.roomId = nil; host.connection.cancel() }
                if let guest = existing.guest { guest.roomId = nil; guest.connection.cancel() }
                rooms.removeValue(forKey: roomId)
            }
            let room = Room(token: token)
            room.host = client
            rooms[roomId] = room
            client.roomId = roomId
            sendJSON(["type": Wire.Signal.roomCreated, "roomId": roomId, "token": token], to: client)

        case Wire.Signal.joinRoom:
            guard let roomId = obj["roomId"] as? String, let room = rooms[roomId] else {
                sendJSON(["type": Wire.Signal.error, "reason": Wire.Reason.roomNotFound], to: client)
                return
            }
            guard (obj["token"] as? String) == room.token else {
                sendJSON(["type": Wire.Signal.error, "reason": Wire.Reason.badToken], to: client)
                return
            }
            guard room.guest == nil else {
                sendJSON(["type": Wire.Signal.error, "reason": Wire.Reason.roomFull], to: client)
                return
            }
            room.guest = client
            client.roomId = roomId
            sendJSON(["type": Wire.Signal.peerJoined, "role": Wire.Role.guest], to: client)
            if let host = room.host {
                sendJSON(["type": Wire.Signal.peerJoined, "role": Wire.Role.host], to: host)
            }

        case Wire.Signal.signal:
            guard let roomId = client.roomId, let room = rooms[roomId] else {
                sendJSON(["type": Wire.Signal.error, "reason": Wire.Reason.notInRoom], to: client)
                return
            }
            if let peer = otherPeer(in: room, of: client), let data = obj["data"] {
                sendJSON(["type": Wire.Signal.signal, "data": data], to: peer)
            }

        case Wire.Signal.message:
            guard let roomId = client.roomId, let room = rooms[roomId] else {
                sendJSON(["type": Wire.Signal.error, "reason": Wire.Reason.notInRoom], to: client)
                return
            }
            if let peer = otherPeer(in: room, of: client), let data = obj[Wire.Key.data] {
                sendJSON([Wire.Key.type: Wire.Signal.message, Wire.Key.data: data], to: peer)
            }

        default:
            sendJSON(["type": Wire.Signal.error, "reason": Wire.Reason.unknownType], to: client)
        }
    }

    private func cleanup(_ client: Client) {
        clients.removeValue(forKey: ObjectIdentifier(client.connection))
        client.connection.cancel()

        guard let roomId = client.roomId, let room = rooms[roomId] else { return }
        let peer = otherPeer(in: room, of: client)
        if room.host === client { room.host = nil }
        if room.guest === client { room.guest = nil }
        if let peer = peer {
            sendJSON(["type": Wire.Signal.peerLeft], to: peer)
        }
        if room.host == nil && room.guest == nil {
            rooms.removeValue(forKey: roomId)
        }
    }

    private func otherPeer(in room: Room, of client: Client) -> Client? {
        if room.host === client { return room.guest }
        if room.guest === client { return room.host }
        return nil
    }

    // MARK: - 发送

    private func sendJSON(_ obj: [String: Any], to client: Client) {
        guard let data = try? JSONSerialization.data(withJSONObject: obj) else { return }
        let meta = NWProtocolWebSocket.Metadata(opcode: .text)
        let context = NWConnection.ContentContext(identifier: "send", metadata: [meta])
        client.connection.send(content: data,
                               contentContext: context,
                               isComplete: true,
                               completion: .contentProcessed { _ in })
    }

    // MARK: - 工具

    private static func genId(bytes: Int) -> String {
        var data = Data(count: bytes)
        _ = data.withUnsafeMutableBytes { ptr in
            SecRandomCopyBytes(kSecRandomDefault, bytes, ptr.baseAddress!)
        }
        return data.map { String(format: "%02x", $0) }.joined()
    }

    private func log(_ line: String) {
        DispatchQueue.main.async { self.onLog?(line) }
    }

    private func emitState(running: Bool, port: UInt16) {
        DispatchQueue.main.async { self.onStateChange?(running, port) }
    }

    private func emitError(_ message: String) {
        DispatchQueue.main.async { self.onError?(message) }
    }
}
